Zend/zend_language_scanner.l

```diff
@@ -1250,6 +1258,69 @@ static void copy_heredoc_label_stack(void *void_heredoc_label)
 	zend_ptr_stack_push(&SCNG(heredoc_label_stack), (void *) new_heredoc_label);
 }
 
+/* Check that { }, [ ], ( ) are nested correctly */
+static void report_bad_nesting(char opening, int opening_lineno, char closing)
+{
+	char buf[256];
+	int  used = 0;
+
+	if (opening == '$') {
+		used = snprintf(buf, sizeof(buf), "Unclosed '${'");
+	} else {
+		used = snprintf(buf, sizeof(buf), "Unclosed '%c'", opening);
+	}
+
+	if (opening_lineno != CG(zend_lineno)) {
+		used += snprintf(buf + used, sizeof(buf) - used, " on line %d", opening_lineno);
+	}
+
+	if (closing) { 	/* 'closing' will be 0 if at end of file */
+		used += snprintf(buf + used, sizeof(buf) - used, " does not match '%c'", closing);
+	}
+
+	zend_throw_exception(zend_ce_parse_error, buf, 0);
+}
+
+static void enter_nesting(char *location)
+{
+	zend_nest_location nest_loc = {location, CG(zend_lineno)};
+	zend_stack_push(&SCNG(nest_location_stack), &nest_loc);
+}
+
+static int exit_nesting(char *location)
+{
+	if (zend_stack_is_empty(&SCNG(nest_location_stack))) {
+		zend_throw_exception_ex(zend_ce_parse_error, 0, "Unmatched '%c'", *location);
+		return -1;
+	}
+
+	zend_nest_location *nest_loc = zend_stack_top(&SCNG(nest_location_stack));
+	char opening = *nest_loc->text;
+	char closing = *location;
+
+	if ((opening == '{' && closing != '}') ||
+	    (opening == '[' && closing != ']') ||
+	    (opening == '(' && closing != ')') ||
+	    (opening == '$' && closing != '}')) { /* for ${ */
+		report_bad_nesting(opening, nest_loc->lineno, closing);
+		return -1;
+	}
+
+	zend_stack_del_top(&SCNG(nest_location_stack));
+	return 0;
+}
+
+static int check_nesting_at_end()
+{
+	if (!zend_stack_is_empty(&SCNG(nest_location_stack))) {
+		zend_nest_location *nest_loc = zend_stack_top(&SCNG(nest_location_stack));
+		report_bad_nesting(*nest_loc->text, nest_loc->lineno, 0);
+		return -1;
```

Comment:
Should `nest_location_stack` be emptied here, so we are ready to start lexing more code using the same `lex_state`? My understanding is that when a parse error is thrown, lexing will stop and probably `shutdown_scanner` will be called, freeing all the data structures allocated in the `lex_state`.

Reply:
Right, clearing the stack is not necessary here, it will get cleaned up when the lexical state is restored.